### PR TITLE
[QA-3113] Move failed tests retries to the testrun end

### PR DESCRIPTION
**Pull request by oiefimenko**
### Summary of Changes
Move failed tests retries execution to the testrun end

### Rationale
Current changes will help to stabilize possible anomalies in fixtures mutations during reruns 

### Relevant JIRA links
https://datarobot.atlassian.net/browse/QA-3113

### TODO after merging

- [ ] Build new package version to artifactory

### CODE REVIEW

File: pytest_rerunfailures.py [hunk 1]
```diff
@@ -71,151 +100,241 @@ def check_options(config):
...
+        Parameters
+        ----------
+        item : _pytest.main.Item
+        """
+        delay = self._get_reruns_delay(item)
+        parallel = hasattr(item.config, 'slaveinput')
+
+        for i in range(reruns):
+            self._invalidate_fixtures(item)
+            time.sleep(delay)
```
Comment (lyapun) [inline pytest_rerunfailures.py]: why do we need this?

**Review by lyapun — COMMENTED** [review]
Do we need to be smart when we invalidate fixtures in this case? I mean, what is the difference if we invalidate all of them, or just those that are used by failed tests? Anyway we will recreate only fixtures that are used by failed tests, right?

Do we have some functional tests for this plugin? We can create like test project inside this repo, in order to unittest this functionality.

File: pytest_rerunfailures.py [hunk 2]
```diff
@@ -71,151 +100,241 @@ def check_options(config):
...
-    """Adapted from https://pytest.org/latest/_modules/_pytest/skipping.html
-    """
-    if report.outcome == 'rerun':
-        return 'rerun', 'R', ('RERUN', {'yellow': True})
+    def pytest_terminal_summary(self, terminalreporter):
+        """
+        Pytest hook
 
+        Handle rerun terminal summary report
+        Adapted from https://pytest.org/latest/_modules/_pytest/skipping.html
```
Comment (andrew-qa) [inline pytest_rerunfailures.py]:
``` 
This page does
 not exist yet 
```

Probably should be replaced by https://docs.pytest.org/en/latest/skipping.html

File: pytest_rerunfailures.py [hunk 3]
```diff
@@ -71,151 +100,241 @@ def check_options(config):
...
+                tr._tw.line(line)
 
-    lines = []
-    for char in tr.reportchars:
-        if char in 'rR':
-            show_rerun(terminalreporter, lines)
+    def _show_rerun(self, terminalreporter, lines):
+        """
+        Format reruned tests to be market as RERUN in output
+        Adapted from https://pytest.org/latest/_modules/_pytest/skipping.html
```
Comment (andrew-qa) [inline pytest_rerunfailures.py]: Same issue with URL

**Review by andrew-qa — COMMENTED** [review]
I agree with @lyapun about functional test for this logic. We should add them here. Also there are problems with links to py.test docs.

File: pytest_rerunfailures.py (hunk 1 above)
Comment (oiefimenko) [reply pytest_rerunfailures.py]: It is adaptation of base rerunfailures plugin - ability to rerun tests with some delay. It is controlled by `--reruns-delay` param

**Comment by oiefimenko** [comment]
Considering unittests - some of the are described in `test_pytest_rerunfailures.py` file. There are more e2e tests on total amount of tests executed, amount of reruns, params pass etc. From this perspective functionality wasn't changed so I didn't introduce new ones.

Looks like pytest don't have any hooks which could be occurs after tests executed but before report publish. So to catch this moment we should track moment when no next tests left (`line 134`) and execute rerun at this point. Because of that we could not be sure that all fixture were already invalidated. 
Second reason I proposed to keep current functionality is that tests scheduled for rerun could be collected from different places in any random order. So in order to not to introduce new flakes I propose to stay with slower but safer approach where all failures will be executed in isolated-like (or at least close to it) way.
And you are right - fixtures will be recreated only for tests scheduled for rerun.

**Comment by lyapun** [comment]
Do you mean that in current implementation we have it in such way:

1. Run all tests. Collect every test that failed.
2. When no tests left - iterate over failed tests and for each test - invalidate this test fixture and run it.

If it's right understanding - I don't understand how it's better than current logic. Still, if all tests will fail - we will recreate fixtures for each of them.

Or I misunderstood your comment?

**Comment by lyapun** [comment]
Also, if we have tests why we haven't setup Jarvis auto testing for this repo? (https://datarobot.atlassian.net/wiki/spaces/JARVIS/pages/304906359/Jarvis+Autotest)

**Comment by oiefimenko** [comment]
Probably I misunderstand task. Reduced fixture invalidation to one time execution.

**Review by lyapun — APPROVED** [review]

**Review by burkestar — COMMENTED** [review]
I didn't review in detail, but it seems good to me.
Can you explain how this was tested to make sure it works as advertised?  Thanks!

**Comment by oiefimenko** [comment]
@burkestar Probably in a simplest way - I created dummy test suits with fixtures and expected failures and check functionality on it. Its besides tests written for package.

**Review by burkestar — APPROVED** [review]
Cool.  I assume we'll need a DataRobot PR to bump the package version and we can do more testing on that branch as needed.

**Pull request merged**